presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayJoin.java

```diff
@@ -0,0 +1,206 @@
+
+public final class ArrayJoin
+        extends ParametricScalar
+{
+    public static final ArrayJoin ARRAY_JOIN = new ArrayJoin();
+    public static final ArrayJoinWithNullReplacement ARRAY_JOIN_WITH_NULL_REPLACEMENT = new ArrayJoinWithNullReplacement();
+
+    private static final TypeSignature VARCHAR_TYPE_SIGNATURE = parseTypeSignature(StandardTypes.VARCHAR);
+    private static final String FUNCTION_NAME = "array_join";
+    private static final String DESCRIPTION = "Concatenates the elements of the given array using a delimiter and an optional string to replace nulls";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(typeParameter("T")), StandardTypes.VARCHAR,
+            ImmutableList.of("array<T>", StandardTypes.VARCHAR),
+            false, false);
+
+    private static class ArrayJoinWithNullReplacement extends ParametricScalar
+    {
+        private final Signature signature = new Signature(FUNCTION_NAME, ImmutableList.of(typeParameter("T")), StandardTypes.VARCHAR,
+                ImmutableList.of("array<T>", StandardTypes.VARCHAR, StandardTypes.VARCHAR),
+                false, false);
+
+        @Override
+        public Signature getSignature()
+        {
+            return signature;
+        }
+
+        @Override
+        public boolean isHidden()
+        {
+            return false;
+        }
+
+        @Override
+        public boolean isDeterministic()
+        {
+            return true;
+        }
+
+        @Override
+        public String getDescription()
+        {
+            return DESCRIPTION;
+        }
+
+        @Override
+        public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+        {
+            return specializeArrayJoin(types, functionRegistry, ImmutableList.of(false, false, false));
+        }
+
+    }
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return DESCRIPTION;
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        return specializeArrayJoin(types, functionRegistry, ImmutableList.of(false, false));
+    }
+
+    private static FunctionInfo specializeArrayJoin(Map<String, Type> types, FunctionRegistry functionRegistry, List<Boolean> nullableArguments)
+    {
+        Type type = types.get("T");
+        TypeSignature arrayType = parameterizedTypeName(StandardTypes.ARRAY, type.getTypeSignature());
+        Signature signature = new Signature(FUNCTION_NAME, VARCHAR_TYPE_SIGNATURE, arrayType, VARCHAR_TYPE_SIGNATURE, VARCHAR_TYPE_SIGNATURE);
```

Comment:
I think you need to take the arguments to the signature, as an argument (or just a boolean indicating that it's the version with replacement). The one that doesn't have the replacement argument should only have two parameters, instead of three